sql/src/main/java/io/crate/analyze/ValuesAwareExpressionAnalyzer.java

```diff
@@ -76,28 +92,55 @@
                                   TransactionContext transactionContext,
                                   Function<ParameterExpression, Symbol> convertParamFunction,
                                   FieldProvider fieldProvider,
-                                  ValuesResolver valuesResolver) {
+                                  ValuesResolver valuesResolver,
+                                  Insert.DuplicateKeyType duplicateKeyType) {
         super(functions, transactionContext, convertParamFunction, fieldProvider, null);
         this.valuesResolver = valuesResolver;
+        this.duplicateKeyType = duplicateKeyType;
     }
 
     @Override
     protected Symbol convertFunctionCall(FunctionCall node, ExpressionAnalysisContext context) {
         List<String> parts = node.getName().getParts();
         if (parts.get(0).equals("values")) {
-            Expression expression = node.getArguments().get(0);
-            Symbol argumentColumn = super.convert(expression, context);
-            if (argumentColumn.valueType().equals(DataTypes.UNDEFINED)) {
-                throw new IllegalArgumentException(
-                    SymbolFormatter.format("Referenced column '%s' in VALUES expression not found", argumentColumn));
+            if (duplicateKeyType != Insert.DuplicateKeyType.ON_DUPLICATE_KEY_UPDATE) {
+                throw new UnsupportedOperationException("Can't use VALUES outside ON DUPLICATE KEY col = VALUES(..)");
             }
-            if (!(argumentColumn instanceof Field)) {
-                throw new IllegalArgumentException(SymbolFormatter.format(
-                    "Argument to VALUES expression must reference a column that " +
-                    "is part of the INSERT statement. %s is invalid", argumentColumn));
-            }
-            return valuesResolver.allocateAndResolve((Field) argumentColumn);
+            Expression expression = node.getArguments().get(0);
+            return convertAllocateAndResolve(expression, valuesResolver, this, context);
         }
         return super.convertFunctionCall(node, context);
     }
+
+    @Override
+    protected Symbol convertQualifiedNameReference(QualifiedNameReference node, ExpressionAnalysisContext context) {
```

Comment:
Couldn't we wrap the `fieldProvder` instead of opening up `convertQualifiedNameReference` in the base class? The `fieldProvider` already provides an extension point exactly to resolve qualified names.

I'd rather lock down the ExpressionAnalyzer so that in the future it's not possible to inherit from it at all (once we got rid of the VALUES hack when we drop MySQL syntax support)

Reply:
Good point. I followed the existing approach via overloads but it's not optimal and introduces changes in the base ExpressionsAnalyzer class. Should be possible to wrap the FieldProvider.